packages/code-studio/src/main/AppMainContainer.jsx

```diff
@@ -203,6 +207,53 @@ export class AppMainContainer extends Component {
     this.widgetListenerRemover?.();
   }
 
+  openNotebookFromURL() {
+    const { match } = this.props;
+    const { notebookPath } = match.params;
+
+    if (notebookPath) {
+      const fileMetadata = {
+        id: `/${notebookPath}`,
+        itemName: `/${notebookPath}`,
+      };
+
+      const { session, sessionConfig } = this.props;
+      const language = sessionConfig.type;
+      const notebookSettings = {
+        value: null,
+        language,
+      };
+
+      const panelConfig = {
+        props: { panelState: { fileMetadata } },
+      };
+
+      const notebookStack = LayoutUtils.getStackForConfig(
+        this.goldenLayout.root,
+        panelConfig
+      );
+
+      if (!notebookStack) {
+        // Open the notebook if it isn't open anywhere
+        this.emitLayoutEvent(
+          NotebookEvent.SELECT_NOTEBOOK,
+          session,
+          language,
+          notebookSettings,
+          fileMetadata,
+          true
+        );
+      } else {
+        // Focus the notebook within its stack if it is already open
+        const notebookPanel = LayoutUtils.getContentItemInStack(
```

Comment:
This doesn't change existing preview tabs into editable notebooks - the tab stays open as a preview.
I think the `NotebookEvent.SELECT_NOTEBOOK` event should be able to handle both new and existing notebook tabs.

Reply:
Calling `NotebookEvent.SELECT_NOTEBOOK` if the notebook is already open causes it to duplicate the notebook. The map of open notebooks isn't populated at the time this is called. The panel is created, but the notebook isn't opened yet which doesn't add it to the map. So `SELECT_NOTEBOOK` finds no open file and says to open a 2nd copy of it

I updated the comment to explain this and was able to set the props on the component before it mounts to make sure it isn't a preview

Reply:
Yea, I think the problem is the `onLayoutInitialized` handler gets called too early before the layout is actually initialized. Might need to rename it to something like 'onConfigLoaded`.

Or... can you try calling `onLayoutInitialized` on the first `stateChanged` event in `DashboardLayout.tsx`? This might work for what we need.